### PR TITLE
Add Grant CLR Pot Handling & Analytics

### PR DESCRIPTION
**Pull request by frankchen07**
##### Description

This PR attempts to address the lack of simple analytics based off of production data, and also the seemingly low or high prediction amounts based on the saturation of the pot.

##### Refers/Fixes

- added an additional clause in the method `predict_clr` that first calculates clr amounts given the pot, modifies the pot, and recalculates predictions based off of the pot. Currently it rounds down, and testing confirms the hypothesis that this doesn't increase projections.
    - the fix might be to round down. I'd like second eyes on this a quick discussion with @thelostone-mc 

- added a method,`analytics_clr` as a simple function that cuts out all donation prediction parameters
    - also added two additional returns in the method`calculate_clr_for_donation`, which carries analytics over from the method `calculate_clr`

---

- EDIT: added 3-type verification handling in clr.py

still needs mods on line `290` and `398` to handle non-boolean string pulls, just the data fetching portion. The calculation logic has been tested with this simple mirroring code:

```
def test(x, y, k2, k1):
    
    # both sms
    if k2 > k1 and all(i in x for i in [k2, k1]):
        print('all in x')
    
    # both bright
    elif k2 > k1 and all(i in y for i in [k2, k1]):
        print('all in y')
    
    # both none
    elif k2 > k1 and not any(i in x+y for i in [k2, k1]):
        print('none verified')
    
    # one bright or sms, one none
    elif k2 > k1 and (((k2 in x+y) and (k1 not in x+y)) or ((k1 in x+y) and (k2 not in x+y))):
        print('one verified one none')
    
    # one bright, one sms
    elif k2 > k1:
        print('mismatched verified')

    return None


x = [1, 2, 3]
y = [4, 5, 6]

test(x, y, 2, 1)  # all in x
test(x, y, 5, 4)  # all in y
test(x, y, 12, 10)  # none verified
test(x, y, 7, 0)  # none verified
test(x, y, 8, 6)  # one verified, one none 
test(x, y, 9, 1)  # one verified, one none
test(x, y, 6, 3)  # mismatched verified
test(x, y, 4, 1)  # mismatched verified
```

---

##### Testing

- `analytics_clr.py` and `estimate_clr.py` successfully run on local

### CODE REVIEW

File: app/grants/management/commands/analytics_clr.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+from dashboard.utils import get_tx_status, has_tx_mined
+from grants.clr import analytics_clr
+from grants.models import Contribution, Grant, GrantCLR
+from marketing.mails import warn_subscription_failed
+
+
+class Command(BaseCommand):
+
+    help = 'calculate CLR estimates for all grants'
```
Comment (thelostone-mc) [inline app/grants/management/commands/analytics_clr.py:33]: ^ could we update this text

File: app/grants/management/commands/analytics_clr.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        clr_pk = options['clr_pk']
+
+        if clr_pk == "all":
+            active_clr_rounds = GrantCLR.objects.filter(is_active=True)
+        else:
+            active_clr_rounds = GrantCLR.objects.filter(pk=clr_pk)
+
+        if active_clr_rounds:
+            for clr_round in active_clr_rounds:
+                print(f"CALCULATING CLR estimates for ROUND: {clr_round.round_num}")
```
Comment (thelostone-mc) [inline app/grants/management/commands/analytics_clr.py:52]: ^ could we update this text

File: app/grants/management/commands/analytics_clr.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                a = analytics_clr(
+                    from_date=timezone.now(),
+                    clr_round=clr_round,
+                    network=network
+                )
+                print(a)
+                print(f"finished CLR estimates for {clr_round.round_num}")
```
Comment (thelostone-mc) [inline app/grants/management/commands/analytics_clr.py:59]: ^ could we update this text

File: app/grants/management/commands/analytics_clr.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+                # TOTAL GRANT
+                # grants = Grant.objects.filter(network=network, hidden=False, active=True, link_to_new_grant=None)
+                # grants = grants.filter(**clr_round.grant_filters)
+
+                # total_clr_distributed = 0
+                # for grant in grants:
+                #     total_clr_distributed += grant.clr_prediction_curve[0][1]
+
+                # print(f'Total CLR allocated for {clr_round.round_num} - {total_clr_distributed}')
```
Comment (thelostone-mc) [inline app/grants/management/commands/analytics_clr.py:69]: ^ Is this something we still need / can be removed ?

File: app/grants/management/commands/estimate_clr.py [hunk 5]
```diff
@@ -50,12 +50,13 @@ def handle(self, *args, **options):
...
             for clr_round in active_clr_rounds:
                 print(f"CALCULATING CLR estimates for ROUND: {clr_round.round_num}")
-                predict_clr(
+                a = predict_clr(
                     save_to_db=True,
                     from_date=timezone.now(),
                     clr_round=clr_round,
                     network=network
                 )
+                print(a)
```
Comment (thelostone-mc) [inline app/grants/management/commands/estimate_clr.py:59]: ^ we can undo the changes made for this file

File: app/grants/clr.py [hunk 6]
```diff
@@ -394,10 +421,42 @@ def predict_clr(save_to_db=False, from_date=None, clr_round=None, network='mainn
...
+    print(f'match hit so far using full pot: {pre_pot}')
+
+    # using clr amount, round total pot up to the closest 10k
+    # I think this needs to be changed to round down
+    total_pot_close = round(pre_pot + 5000, -4) 
+
+    print(f'using {total_pot_close} to calculate preds based on full pot') 
+
+    # # when not using our pot mechanisms
+    # total_pot_close = total_pot
```
Comment (thelostone-mc) [inline app/grants/clr.py:458]: What are you thoughts on shifting this over to
https://github.com/gitcoinco/web/blob/master/app/grants/clr.py#L208

That is where our normalization code resides !
Modifying the total_pot to use `total_pot_close` would mess up the value of `CLR_PERCENTAGE_DISTRIBUTED`

We could do something like

```
    if bigtot >= total_pot: # saturation reached
        # print(f'saturation reached. Total Pot: ${total_pot} | Total Allocated ${bigtot}. Normalizing')
        CLR_PERCENTAGE_DISTRIBUTED = 100
        for t in totals:
            t['clr_amount'] = ((t['clr_amount'] / bigtot) * total_pot)
    else:
        CLR_PERCENTAGE_DISTRIBUTED =  (bigtot / total_pot) * 100
        # your code starts here
        where you can define a `temp_total_pot` is X% higher than `bigtot`
       and then do something like 
        for t in totals:
            t['clr_amount'] = ((t['clr_amount'] / bigtot) * temp_total_pot)


```

**Review by thelostone-mc — CHANGES_REQUESTED** [review]
Left a few comments ! 
The cleanup is a nice add <3

File: app/grants/management/commands/analytics_clr.py (hunk 1 above)
Comment (frankchen07) [reply app/grants/management/commands/analytics_clr.py:33]: done

File: app/grants/management/commands/analytics_clr.py (hunk 2 above)
Comment (frankchen07) [reply app/grants/management/commands/analytics_clr.py:52]: done

File: app/grants/management/commands/analytics_clr.py (hunk 3 above)
Comment (frankchen07) [reply app/grants/management/commands/analytics_clr.py:59]: done

File: app/grants/management/commands/analytics_clr.py (hunk 4 above)
Comment (frankchen07) [reply app/grants/management/commands/analytics_clr.py:69]: removed

File: app/grants/management/commands/estimate_clr.py (hunk 5 above)
Comment (frankchen07) [reply app/grants/management/commands/estimate_clr.py:59]: done

File: app/grants/management/commands/estimate_clr.py [hunk 7]
```diff
@@ -50,24 +50,14 @@ def handle(self, *args, **options):
         if active_clr_rounds:
             for clr_round in active_clr_rounds:
                 print(f"CALCULATING CLR estimates for ROUND: {clr_round.round_num}")
-                predict_clr(
+                a = predict_clr(
```
Comment (thelostone-mc) [inline app/grants/management/commands/estimate_clr.py:53]: we can remove this

File: app/grants/management/commands/estimate_clr.py [hunk 8]
```diff
@@ -50,24 +50,14 @@ def handle(self, *args, **options):
...
-                # TOTAL GRANT
-                # grants = Grant.objects.filter(network=network, hidden=False, active=True, link_to_new_grant=None)
-                # grants = grants.filter(**clr_round.grant_filters)
-
-                # total_clr_distributed = 0
-                # for grant in grants:
-                #     total_clr_distributed += grant.clr_prediction_curve[0][1]
-
-                # print(f'Total CLR allocated for {clr_round.round_num} - {total_clr_distributed}')
-
```
Comment (thelostone-mc) [inline app/grants/management/commands/estimate_clr.py:70]: ^ Could we leave this in for testing reasons

File: app/grants/clr.py [hunk 9]
```diff
@@ -394,6 +449,8 @@ def predict_clr(save_to_db=False, from_date=None, clr_round=None, network='mainn
     v_threshold = float(clr_round.verified_threshold)
     uv_threshold = float(clr_round.unverified_threshold)
 
+    print(total_pot)
```
Comment (thelostone-mc) [inline app/grants/clr.py:452]: remove!

File: app/grants/clr.py [hunk 10]
```diff
@@ -384,6 +407,38 @@ def populate_data_for_clr(grants, contributions, phantom_funding_profiles, clr_r
 
 
 
+def analytics_clr(from_date=None, clr_round=None, network='mainnet'):
```
Comment (thelostone-mc) [inline app/grants/clr.py:410]: Do we wanna move this into it's own file ?

File: app/grants/management/commands/estimate_clr.py (hunk 7 above)
Comment (frankchen07) [reply app/grants/management/commands/estimate_clr.py:53]: done

File: app/grants/management/commands/estimate_clr.py (hunk 8 above)
Comment (frankchen07) [reply app/grants/management/commands/estimate_clr.py:70]: yas done

File: app/grants/clr.py (hunk 9 above)
Comment (frankchen07) [reply app/grants/clr.py:452]: done

File: app/grants/clr.py (hunk 10 above)
Comment (frankchen07) [reply app/grants/clr.py:410]: I just left it there because if we moved it into its own file...I'm assuming we would have to port over alot of the same methods used in the `clr.py` file, no?

Comment (thelostone-mc) [reply app/grants/clr.py:410]: Oh you can still import those methods into your analytics file !
If you need xyz method
import xyz from app/grants/clr.py sort of thing

Comment (frankchen07) [reply app/grants/clr.py]: done. So, maybe a stupid question. How come in `analytics_clr.py` I have to import ` fetch_data`, `populate_data_for_clr`, and `calculate_clr_for_donation` but you don't have to do so in `estimate_clr.py`, even though `predict_clr` uses the same methods?

Comment (thelostone-mc) [reply app/grants/clr.py]: Oh no you wouldn't have to do that 🙌

**Comment by thelostone-mc** [comment]
LOOKS SOLIDDDD :D I'll add the final pieces

**Review by thelostone-mc — APPROVED** [review]

**Comment by frankchen07** [comment]
looks good, I had an additional question. I thought I asked it but I can't find it:

for the analytics script, how come I have to import those three methods for `analytics_clr` (fetching, populate data, and calculate clr for donation) in `analytics_clr.py` but for `estimate_clr.py` we only have to import `predict_clr`? Shouldn't we have to import those methods in `estimate_clr.py` too?

File: app/grants/clr.py [hunk 11]
```diff
@@ -350,10 +379,15 @@ def populate_data_for_clr(grants, contributions, phantom_funding_profiles, clr_r
...
-        verified_profile_ids = [ele.pk for ele in contribs if ele.profile_for_clr.sms_verification]
-        verified_phantom_funding_profile_ids = [ele.profile_id for ele in grant_phantom_funding_profiles if ele.profile.sms_verification]
-        verified_profile = list(set(verified_profile_ids + verified_phantom_funding_profile_ids))
+        # SMS verified profiles
+        sms_verified_profile_ids = [ele.pk for ele in contribs if ele.profile_for_clr.sms_verification]
+        sms_verified_phantom_funding_profile_ids = [ele.profile_id for ele in grant_phantom_funding_profiles if ele.profile.sms_verification]
+        sms_verified_profile = list(set(sms_verified_profile_ids + sms_verified_phantom_funding_profile_ids))
+
+        # BrightID verified profiles
+        brightid_verified_profile_ids = [ele.pk for ele in contribs if ele.profile_for_clr.is_brightid_verified]
```
Comment (thelostone-mc) [inline app/grants/clr.py:388]: @frankchen07 mapping this to https://github.com/gitcoinco/web/pull/7333/files#r485161156

Comment (frankchen07) [reply app/grants/clr.py:388]: ossom, LGTM, makes sense.

- we have two lists
- clr calculations takes both
- we calculate multipliers based on set inclusions

**Comment by thelostone-mc** [comment]
Closing as this will get in via 
https://github.com/ScopeLift/gitcoin-web/pull/86
